File: Python/tests/SD_control_test.py
import pygame

class JoystickController:
    def __init__(self):
        pygame.init()
        pygame.joystick.init()  # Initialize the joystick module
        self.joystick = pygame.joystick.Joystick(0)  # Initialize the first joystick
        self.joystick.init()
        self.prev_x = None
        self.prev_y = None
        self.prev_speed = None

    def get_joystick_position_and_speed(self):
        pygame.event.pump()  # Process event queue

        # Get the position of the joystick
        x_axis = self.joystick.get_axis(0)
        y_axis = self.joystick.get_axis(1)

        # Apply a deadzone to the joystick to prevent drift for X-axis
        if abs(x_axis) < 0.3:
            x_axis = 0
        else:
            # Invert the x_axis value to correct the turning direction
            x_axis = round(-x_axis, 4)  # Invert the direction of the turn

        # Adjust Y-axis to be -1, 0, or 1 based on its direction or neutral position
        if abs(y_axis) < 0.3 or abs(x_axis) == 1 or abs(x_axis) == -1:
            y_axis = 0
        else:
            y_axis = round(y_axis / abs(y_axis))  # This will result in -1 or 1
            
        if 1 > x_axis > 0 and y_axis == 0:
            x_axis = 1
            
        if -1 < x_axis < 0 and y_axis == 0:
            x_axis = -1

        if x_axis == 1 and y_axis == 0:
            x_axis = -1
        elif x_axis == -1 and y_axis == 0:
            x_axis = 1

        # Get the value of the right trigger (RT on Xbox controller)
        trigger_value = self.joystick.get_axis(5)  # Adjust if necessary for your controller
        corrected_trigger_value = (trigger_value + 1) / 2  # Normalize trigger value from -1 to 1 to 0 to 1

        # Speed handling remains the same
        if corrected_trigger_value < 0.1:
            speed = 0
        else:
            speed = round(0.4 + ((trigger_value + 1) / 2) * 0.6, 4)  # Map trigger value to range 0.4 to 1

        return x_axis, y_axis, speed
    
    def get_input(self):
        x, y, speed = self.get_joystick_position_and_speed()
        return x, y, speed
    
    def start(self):
        pass

    def close(self):
        pygame.quit()

def main():
    controller = JoystickController()
    try:
        while True:
            x, y, speed = controller.get_input()
            if x != controller.prev_x or y != controller.prev_y or speed != controller.prev_speed:
                print(f"Joystick Position: ({x}, {y}), Trigger Position: {speed}")
                controller.prev_x = x
                controller.prev_y = y
                controller.prev_speed = speed
    except KeyboardInterrupt:
        controller.close()

if __name__ == "__main__":
    main()

File: Cpp/tests/SD_control_test.cpp
#include <iostream>
#include <SDL2/SDL.h>

class JoystickController {
private:
    SDL_Joystick* joystick;
    int prev_x;
    int prev_y;
    int prev_speed;

public:
    JoystickController() {
        SDL_Init(SDL_INIT_JOYSTICK);
        joystick = SDL_JoystickOpen(0);
        prev_x = 0;
        prev_y = 0;
        prev_speed = 0;
    }

    std::tuple<int, int, int> get_joystick_position_and_speed() {
        SDL_Event event;
        SDL_PumpEvents();

        // Get the position of the joystick
        int x_axis = SDL_JoystickGetAxis(joystick, 0);
        int y_axis = SDL_JoystickGetAxis(joystick, 1);

        // Apply deadzone
        if (abs(x_axis) < 3000) {
            x_axis = 0;
        }

        if (abs(y_axis) < 3000) {
            y_axis = 0;
        }

        // Get the value of the right trigger
        int trigger_value = SDL_JoystickGetAxis(joystick, 5);
        int corrected_trigger_value = (trigger_value + 32767) / 2;

        int speed;
        if (corrected_trigger_value < 10000) {
            speed = 0;
        } else {
            speed = 0.4 + ((trigger_value + 32767) / 65535.0) * 0.6;
        }

        return std::make_tuple(x_axis, y_axis, speed);
    }

    void start() {
        // Not needed for this example
    }

    void close() {
        SDL_JoystickClose(joystick);
        SDL_Quit();
    }
};

int main() {
    JoystickController controller;

    try {
        while (true) {
            auto [x, y, speed] = controller.get_joystick_position_and_speed();
            if (x != controller.prev_x || y != controller.prev_y || speed != controller.prev_speed) {
                std::cout << "Joystick Position: (" << x << ", " << y << "), Trigger Position: " << speed << std::endl;
                controller.prev_x = x;
                controller.prev_y = y;
                controller.prev_speed = speed;
            }
        }
    } catch (...) {
        controller.close();
    }

    return 0;
}
